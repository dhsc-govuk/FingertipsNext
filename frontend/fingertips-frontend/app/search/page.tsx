import { SearchForm } from '@/components/forms/SearchForm';

export default async function Page(
  props: Readonly<{
    searchParams?: Promise<{
      indicator?: string;
    }>;
  }>
) {
  const searchParams = await props.searchParams;
  const indicator = searchParams?.indicator ?? '';

  return <SearchForm indicator={indicator} />;
}
