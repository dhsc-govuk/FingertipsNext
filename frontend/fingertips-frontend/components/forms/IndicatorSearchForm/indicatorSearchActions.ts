'use server';

import { z } from 'zod';
import { redirect, RedirectType } from 'next/navigation';
import {
  SearchParams,
  SearchStateManager,
  SearchStateParams,
} from '@/lib/searchStateManager';

const $IndicatorSearchFormSchema = z
  .object({
    searchState: z.string(),
    indicator: z.string(),
  })
  .refine(
    (data) => {
      const stateParsed = JSON.parse(data.searchState);
      if (
        data.indicator.trim().length > 0 ||
        stateParsed[SearchParams.AreasSelected]?.length > 0
      ) {
        return true;
      }
      return false;
    },
    {
      message: 'Please enter an indicator id',
    }
  );

export type State = {
  errors?: {
    indicator?: string[];
  };
  message?: string | null;
};

export type IndicatorSearchForm = z.infer<typeof $IndicatorSearchFormSchema>;

export type IndicatorSearchFormState<T = IndicatorSearchForm> = State & T;

export async function searchIndicator(
  prevState: IndicatorSearchFormState,
  formData: FormData
): Promise<IndicatorSearchFormState> {
  const validatedFields = $IndicatorSearchFormSchema.safeParse({
    searchState: formData.get('searchState'),
    indicator: formData.get('indicator'),
  });

  const searchStateParsed: SearchStateParams = JSON.parse(
    formData.get('searchState')?.toString() ?? `{}`
  );
  const searchStateManager = SearchStateManager.initialise({
    ...searchStateParsed,
  });

  if (!validatedFields.success) {
    return {
      searchState: formData.get('searchState')?.toString() ?? '',
      indicator: formData.get('indicator')?.toString().trim() ?? '',
      errors: validatedFields.error.flatten().fieldErrors,
      message: 'Please enter a subject',
    };
  }

  const { indicator } = validatedFields.data;

  searchStateManager.addParamValueToState(
    SearchParams.SearchedIndicator,
    indicator
  );

  redirect(searchStateManager.generatePath('/results'), RedirectType.push);
}
