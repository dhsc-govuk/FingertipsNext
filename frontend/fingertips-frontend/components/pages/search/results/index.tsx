'use client';

import { BackLink, H1, Paragraph } from 'govuk-react';

export function SearchResults({ indicator }: Readonly<{ indicator: string }>) {
  return (
    <>
      <BackLink href={`/search?indicator=${indicator}`} />
      <H1>Search results</H1>
      <Paragraph>{`You searched for indicator "**${indicator}**"`}</Paragraph>
    </>
  );
}
