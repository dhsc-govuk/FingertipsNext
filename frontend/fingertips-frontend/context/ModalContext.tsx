import {
  createContext,
  FC,
  ReactNode,
  useContext,
  useMemo,
  useState,
} from 'react';
import { ModalPane } from '@/components/molecules/ModalPane/ModalPane';

export interface Modal {
  content?: ReactNode;
  title?: string;
}

export const ModalContext = createContext<{
  modal: Modal;
  setModal: (newPreview: Modal) => void;
}>({ modal: {}, setModal: () => undefined });

export const useModal = () => {
  const { modal, setModal } = useContext(ModalContext);
  const { content, title } = modal;
  return { modal, setModal, content, title };
};

interface ModalProviderProps {
  children: ReactNode;
}

export const ModalProvider: FC<ModalProviderProps> = ({ children }) => {
  const [modal, setModal] = useState<Modal>({});
  const providerValue = useMemo(
    () => ({
      modal,
      setModal,
    }),
    [modal]
  );

  return (
    <ModalContext.Provider value={providerValue}>
      {children}
      <ModalPane />
    </ModalContext.Provider>
  );
};
