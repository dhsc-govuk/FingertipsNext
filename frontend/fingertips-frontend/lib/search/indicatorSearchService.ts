import {
  AzureKeyCredential,
  SearchClient,
  SearchOptions,
} from '@azure/search-documents';

import {
  IIndicatorSearchService,
  INDICATOR_SEARCH_INDEX_NAME,
  IndicatorDocument,
  RawIndicatorDocument,
} from './searchTypes';
import { IndicatorMapper } from './indicatorMapper';

export class IndicatorSearchService implements IIndicatorSearchService {
  private readonly searchClient: SearchClient<RawIndicatorDocument>;
  private readonly mapper: IndicatorMapper;

  constructor(fingertipsAzureAiSearchUrl: string, apiKey: string) {
    const indexName = INDICATOR_SEARCH_INDEX_NAME;
    const credentials = new AzureKeyCredential(apiKey);

    this.searchClient = new SearchClient<RawIndicatorDocument>(
      fingertipsAzureAiSearchUrl,
      indexName,
      credentials
    );
    this.mapper = new IndicatorMapper();
  }

  async searchWith(
    searchTerm: string,
    isEnglandSelectedAsGroup: boolean,
    areaCodes?: string[]
  ): Promise<IndicatorDocument[]> {
    // If the search is entirely numeric, we prioritise searcing for the exact id in the indicator id field,
    // but include a wildcard search as some indicator titles include numbers (ie '65').
    //
    // If the search is not numeric, we prioritise matching the full term against the indicator name
    // and allow fuzzy matching against each word in the search term
    const buildIndicatorIdSearchString = (searchTerm: string): string => {
      return `indicatorID:${searchTerm}^2 ${searchTerm}`;
    };

    const buildTextSearchString = (searchTerm: string): string => {
      const fuzzySearchTerms = searchTerm
        .split(/\s+/)
        .map((subString) => {
          return subString.length > 5 ? `${subString}~` : subString;
        })
        .join(' ')
        .trim();
      return `indicatorName:"${searchTerm}"^2 ${fuzzySearchTerms}`;
    };

    const isPossibleIndicatorId = !isNaN(Number(searchTerm));
    const query = isPossibleIndicatorId
      ? buildIndicatorIdSearchString(searchTerm)
      : buildTextSearchString(searchTerm);

    // This creates an AI Search filter string which should look like
    //  associatedAreaCodes/any(a: a eq 'E09000023' or a eq 'E09000013' or a eq 'E09000025')
    const formatFilterString = (areaCodes: string[]) => {
      if (areaCodes.length == 0) return undefined;
      const areaCodeEqualityStrings = areaCodes.map((a) => `a eq '${a}'`);
      return `associatedAreaCodes/any(a: ${areaCodeEqualityStrings.join(' or ')})`;
    };

    const searchOptions: SearchOptions<RawIndicatorDocument> = {
      queryType: 'full',
      includeTotalCount: true,
      top: 100,
      filter: areaCodes ? formatFilterString(areaCodes) : undefined,
    };

    const searchResponse = await this.searchClient.search(query, searchOptions);
    const results: RawIndicatorDocument[] = [];

    for await (const result of searchResponse.results) {
      results.push(result.document as RawIndicatorDocument);
    }

    return this.mapper.toEntities(
      results,
      areaCodes ?? [],
      isEnglandSelectedAsGroup
    );
  }

  async getIndicator(
    indicatorId: string
  ): Promise<IndicatorDocument | undefined> {
    let rawIndicatorDocument: RawIndicatorDocument;
    try {
      rawIndicatorDocument = (await this.searchClient.getDocument(
        indicatorId
      )) as RawIndicatorDocument;
    } catch (e) {
      console.error(
        `error getting single indicator from ai search: ${(<Error>e).message}`
      );
      return undefined;
    }

    return this.mapper.toEntity(rawIndicatorDocument, [], false);
  }
}
