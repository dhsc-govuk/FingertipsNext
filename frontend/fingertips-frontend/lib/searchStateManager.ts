import { asArray } from './pageHelpers';

export enum SearchParams {
  SearchedIndicator = 'si',
  IndicatorsSelected = 'is',
  AreasSelected = 'as',
  AreaTypeSelected = 'ats',
  GroupTypeSelected = 'gts',
  ConfidenceIntervalSelected = 'cis',
}

export type SearchParamKeys = `${SearchParams}`;

const multiValueParams = [
  SearchParams.IndicatorsSelected as string,
  SearchParams.AreasSelected as string,
  SearchParams.ConfidenceIntervalSelected as string,
];

export type SearchStateParams = {
  [SearchParams.SearchedIndicator]?: string;
  [SearchParams.IndicatorsSelected]?: string[];
  [SearchParams.AreasSelected]?: string[];
  [SearchParams.AreaTypeSelected]?: string;
  [SearchParams.GroupTypeSelected]?: string;
  [SearchParams.ConfidenceIntervalSelected]?: string[];
};

const isMultiValueTypeParam = (searchParamKey: SearchParamKeys) =>
  multiValueParams.includes(searchParamKey);

export class SearchStateManager {
  private searchState: SearchStateParams;
  private searchStateParams: URLSearchParams;

  constructor(searchState: SearchStateParams = {}) {
    this.searchState = {};
    this.searchStateParams = new URLSearchParams();

    this.searchState = searchState;
  }

  private constructPath(path: string) {
    if (this.searchStateParams.size === 0) {
      return path;
    }
    return `${path}?${this.searchStateParams.toString()}`;
  }

  private addParamToPath(searchParamKey: SearchParamKeys) {
    const paramsToAdd = asArray(this.searchState[searchParamKey]);

    if (paramsToAdd.length > 0) {
      paramsToAdd.forEach((paramToAdd) => {
        this.searchStateParams.append(searchParamKey, paramToAdd);
      });
    }
  }

  public addParamValueToState(
    searchParamKey: SearchParamKeys,
    paramToAdd: string
  ) {
    if (isMultiValueTypeParam(searchParamKey)) {
      const currentParamState = this.searchState[searchParamKey];

      const currentParamStateAsArray = asArray(currentParamState);
      const doesParamExist = currentParamStateAsArray.find(
        (currentParam) => currentParam === paramToAdd
      );

      if (!doesParamExist) {
        const newState: SearchStateParams = {
          ...this.searchState,
          [searchParamKey]: [...currentParamStateAsArray, paramToAdd],
        };

        this.searchState = newState;
      }
    } else {
      const newState: SearchStateParams = {
        ...this.searchState,
        [searchParamKey]: paramToAdd,
      };

      this.searchState = newState;
    }
  }

  public removeParamValueFromState(
    searchParamKey: SearchParamKeys,
    paramToRemove?: string
  ) {
    if (isMultiValueTypeParam(searchParamKey)) {
      const currentParamState = asArray(this.searchState[searchParamKey]);

      const newParamState = currentParamState?.filter((paramValue) => {
        return paramValue !== paramToRemove;
      });

      const newState: SearchStateParams = {
        ...this.searchState,
        [searchParamKey]: newParamState,
      };

      this.searchState = newState;
    } else {
      const newState: SearchStateParams = {
        ...this.searchState,
        [searchParamKey]: undefined,
      };

      this.searchState = newState;
    }
  }

  public removeAllParamFromState(searchParamKey: SearchParamKeys) {
    this.searchState[searchParamKey] = undefined;
  }

  public static setStateFromParams(params: URLSearchParams) {
    const newState = Object.values(SearchParams).reduce<SearchStateParams>(
      (state, searchParamKey) => {
        if (isMultiValueTypeParam(searchParamKey)) {
          const paramValues = params.getAll(searchParamKey);

          if (paramValues.length > 0) {
            const newState: SearchStateParams = {
              ...state,
              [searchParamKey]: paramValues,
            };

            return newState;
          }
        } else {
          const paramValue = params.get(searchParamKey);

          if (paramValue) {
            const newState: SearchStateParams = {
              ...state,
              [searchParamKey]: paramValue,
            };

            return newState;
          }
        }
        return state;
      },
      {}
    );

    const searchStateManager = new SearchStateManager(newState);
    return searchStateManager;
  }

  public static setStateFromSearchStateParams(params?: SearchStateParams) {
    if (params) {
      const newState = Object.values(SearchParams).reduce<SearchStateParams>(
        (state, searchParamKey) => {
          if (isMultiValueTypeParam(searchParamKey)) {
            const paramValues = asArray(params[searchParamKey]);

            if (paramValues.length > 0) {
              const newState: SearchStateParams = {
                ...state,
                [searchParamKey]: paramValues,
              };

              return newState;
            }
          } else {
            const paramValue = params[searchParamKey];

            if (paramValue) {
              const newState: SearchStateParams = {
                ...state,
                [searchParamKey]: paramValue,
              };

              return newState;
            }
          }
          return state;
        },
        {}
      );

      const searchStateManager = new SearchStateManager(newState);
      return searchStateManager;
    }
    return new SearchStateManager();
  }

  public getSearchState() {
    return this.searchState;
  }

  public generatePath(path: string) {
    this.searchStateParams = new URLSearchParams();

    Object.values(SearchParams).forEach((searchParamKey) => {
      this.addParamToPath(searchParamKey);
    });

    return this.constructPath(path);
  }
}
