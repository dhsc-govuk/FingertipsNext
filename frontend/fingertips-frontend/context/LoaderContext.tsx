'use client';

import React, {
  ReactNode,
  createContext,
  useContext,
  useMemo,
  useState,
  useEffect,
  Suspense,
  Dispatch,
  SetStateAction,
} from 'react';
import { usePathname } from 'next/navigation';
import { LoadingBox } from 'govuk-react';
import { useSearchStateParams } from '@/components/hooks/useSearchStateParams';
import { SearchStateParams } from '@/lib/searchStateManager';

export type LoaderContext = {
  getIsLoading: () => boolean;
  setIsLoading: (loading: boolean) => void;
};

type LoaderContextProvider = {
  children: ReactNode;
};

const LoaderContext = createContext<LoaderContext | undefined>(undefined);

export const LoaderProvider: React.FC<LoaderContextProvider> = ({
  children,
}) => {
  const pathname = usePathname();
  const [searchState, setSearchState] = useState<SearchStateParams>({});

  const [isLoading, setIsLoading] = useState<boolean>(true);

  const contextValue: LoaderContext = useMemo(
    () => ({
      getIsLoading: () => isLoading,
      setIsLoading,
    }),
    [isLoading, setIsLoading]
  );

  useEffect(() => {
    setIsLoading(false);
  }, [setIsLoading, pathname, searchState]);

  return (
    <LoaderContext.Provider value={contextValue}>
      <Suspense>
        <SuspendedUseSearchStateParams setSearchState={setSearchState} />
      </Suspense>
      <LoadingBox loading={isLoading} timeIn={200} timeOut={200}>
        {children}
      </LoadingBox>
    </LoaderContext.Provider>
  );
};

export const useLoadingState = () => {
  const context = useContext(LoaderContext);

  if (!context) {
    throw new Error('useLoadingState must be used within a LoaderProvider');
  }

  return context;
};

// Next has a very peculiar quirk where useSearchState cannot be called server
// side, which works fine during dev but build will fail.
// Here we're creating a component which will not render server side by
// suspending it until client mounts at which point it can set the loading
// context value, all this is more complicated because loader context is in a
// layout.
function SuspendedUseSearchStateParams({
  setSearchState,
}: Readonly<{ setSearchState: Dispatch<SetStateAction<SearchStateParams>> }>) {
  const searchParams = useSearchStateParams();
  useEffect(() => {
    setSearchState(searchParams);
  }, [searchParams, setSearchState]);

  return null;
}
