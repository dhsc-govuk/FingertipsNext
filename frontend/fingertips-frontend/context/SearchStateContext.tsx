'use client';

import React, { ReactNode, createContext, useContext, useState } from 'react';
import { SearchStateParams } from '@/lib/searchStateManager';

export type SearchStateContext = {
  getSearchState: () => SearchStateParams;
  setSearchState: (state: SearchStateParams) => void;
};

type SearchStateContextProvider = {
  children: ReactNode;
};

const SearchStateContext = createContext<SearchStateContext | undefined>(
  undefined
);

export const SearchStateProvider: React.FC<SearchStateContextProvider> = ({
  children,
}) => {
  const [searchState, setSearchState] = useState<SearchStateParams>({});

  const contextValue: SearchStateContext = {
    getSearchState: () => searchState,
    setSearchState,
  };

  return (
    <SearchStateContext.Provider value={contextValue}>
      {children}
    </SearchStateContext.Provider>
  );
};

export const useSearchState = () => {
  const context = useContext(SearchStateContext);

  if (!context) {
    throw new Error('useLoadingState must be used within a LoaderProvider');
  }

  return context;
};
