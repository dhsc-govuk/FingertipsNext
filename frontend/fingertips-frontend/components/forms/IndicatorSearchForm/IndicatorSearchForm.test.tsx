import { expect } from '@jest/globals';
import { render, screen } from '@testing-library/react';
import { IndicatorSearchFormState } from './indicatorSearchActions';
import { IndicatorSearchForm } from '.';
import { SearchParams, SearchStateParams } from '@/lib/searchStateManager';
import { LoaderContext } from '@/context/LoaderContext';
import userEvent from '@testing-library/user-event';
import { INDICATOR_SEARCH_MAX_CHARACTERS } from '@/lib/search/indicatorSearchService';

jest.mock('react', () => {
  const originalModule = jest.requireActual('react');

  return {
    ...originalModule,
    useActionState: jest
      .fn()
      .mockImplementation(
        (
          _: (
            formState: IndicatorSearchFormState,
            formData: FormData
          ) => Promise<IndicatorSearchFormState>,
          initialState: IndicatorSearchFormState
        ) => [initialState, '/action']
      ),
  };
});

const mockSetIsLoading = jest.fn();
const mockLoaderContext: LoaderContext = {
  getIsLoading: jest.fn(),
  setIsLoading: mockSetIsLoading,
};
jest.mock('@/context/LoaderContext', () => {
  return {
    useLoadingState: () => mockLoaderContext,
  };
});

const mockIndicatorValue = 'test value';
const searchState: SearchStateParams = {
  [SearchParams.SearchedIndicator]: mockIndicatorValue,
};
jest.mock('@/components/hooks/useSearchStateParams', () => ({
  useSearchStateParams: () => searchState,
}));

const initialState: IndicatorSearchFormState = {
  searchState: JSON.stringify(searchState),
  indicator: mockIndicatorValue,
  message: null,
  errors: {},
};

it('snapshot test - renders the form', () => {
  const container = render(
    <IndicatorSearchForm indicatorSearchFormState={initialState} />
  );

  expect(container.asFragment()).toMatchSnapshot();
});

it('should have an input field to input the indicatorId', () => {
  render(<IndicatorSearchForm indicatorSearchFormState={initialState} />);

  expect(screen.getByRole('searchbox')).toBeInTheDocument();
});

it('should set the input field with indicator value from the form state', () => {
  render(<IndicatorSearchForm indicatorSearchFormState={initialState} />);

  expect(screen.getByRole('searchbox')).toHaveValue(mockIndicatorValue);
});

it('should call setIsLoading with true when the search button is clicked', async () => {
  render(<IndicatorSearchForm indicatorSearchFormState={initialState} />);

  const user = userEvent.setup();
  await user.click(screen.getByRole('button'));

  expect(mockSetIsLoading).toHaveBeenCalledWith(true);
});

it('should display no message when there is no error', () => {
  render(<IndicatorSearchForm indicatorSearchFormState={initialState} />);

  expect(
    screen.queryByTestId('indicator-search-form-error')
  ).not.toBeInTheDocument();
});

it('should display an error message when there is an error', () => {
  const searchFormState: IndicatorSearchFormState = {
    ...initialState,
    message: 'error message',
    errors: {},
  };
  render(<IndicatorSearchForm indicatorSearchFormState={searchFormState} />);

  expect(
    screen.queryByTestId('indicator-search-form-error')
  ).toBeInTheDocument();
});

it('should not display character if indicator is under the 75% threshold', () => {
  const searchFormState: IndicatorSearchFormState = {
    ...initialState,
    indicator: 'A'.repeat(INDICATOR_SEARCH_MAX_CHARACTERS * 0.75),
  };
  render(<IndicatorSearchForm indicatorSearchFormState={searchFormState} />);

  expect(
    screen.queryByText('You have 50 characters remaining.', { exact: false })
  ).not.toBeInTheDocument();
});

it('should display character count if indicator is over the 75% threshold', () => {
  const searchFormState: IndicatorSearchFormState = {
    ...initialState,
    indicator: 'A'.repeat(INDICATOR_SEARCH_MAX_CHARACTERS * 0.75 + 1),
  };
  render(<IndicatorSearchForm indicatorSearchFormState={searchFormState} />);

  expect(
    screen.getByText('You have 49 characters remaining.')
  ).toBeInTheDocument();
});

it('should update character count on change', async () => {
  const searchFormState: IndicatorSearchFormState = {
    ...initialState,
    indicator: 'A'.repeat(INDICATOR_SEARCH_MAX_CHARACTERS * 0.75),
  };
  render(<IndicatorSearchForm indicatorSearchFormState={searchFormState} />);

  const user = userEvent.setup();

  await user.type(screen.getByRole('searchbox'), 'A');

  expect(
    screen.getByText('You have 49 characters remaining.')
  ).toBeInTheDocument();
});
