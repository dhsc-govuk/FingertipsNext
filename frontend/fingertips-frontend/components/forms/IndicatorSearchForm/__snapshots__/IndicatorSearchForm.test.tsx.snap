// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test - renders the form 1`] = `
<DocumentFragment>
  .c2 {
  margin: 0;
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  margin-bottom: 15px;
}

.c2 > p {
  margin: 0;
}

.c2 > p > code {
  padding: 0.2em 0.4em;
  margin: 0;
  font-size: 85%;
  background-color: rgba(27,31,35,0.05);
  border-radius: 3px;
}

.c2 > pre {
  padding: 16px;
  overflow: auto;
  font-size: 85%;
  line-height: 1.45;
  background-color: #f6f8fa;
  border-radius: 3px;
}

.c2 > pre > code {
  display: inline;
  padding: 0;
  margin: 0;
  border: 0;
  overflow: visible;
  line-height: inherit;
  word-wrap: normal;
}

.c1 {
  margin-bottom: 20px;
  padding-left: 15px;
  border-left: 5px solid #d4351c;
}

.c0 .c0:last-of-type {
  margin-bottom: 0;
}

.c1.c1 .c1 {
  padding: 0;
  border: 0;
}

.c11 {
  display: block;
}

.c9 {
  width: 100%;
  height: 40px;
  padding: 6px;
  margin: 0;
  border: solid 1px #bfc1c3;
  border-right: 0;
  box-sizing: border-box;
  font-family: "nta",Arial,sans-serif;
  font-weight: 400;
  text-transform: none;
  font-size: 16px;
  line-height: 1.75;
  background: #ffffff;
  border-radius: 0;
  -webkit-appearance: none;
}

.c9:focus {
  margin-right: 3px;
  outline: 3px solid #ffdd00;
  outline-offset: 0;
}

.c9:focus ~ button {
  width: 46px;
}

.c10 {
  background-color: #2b8cc4;
  border: 0;
  display: block;
  color: #ffffff;
  position: relative;
  padding: 10px;
  width: 45px;
  height: 40px;
  background-repeat: no-repeat;
  background-position: 2px 50%;
  text-indent: -999em;
  overflow: hidden;
}

.c10:focus {
  outline: 3px solid #ffdd00;
  outline-offset: 0;
}

.c7 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  background: #ffffff;
  margin-bottom: 0;
}

.c8 {
  margin-bottom: 20px;
}

.c3 {
  padding-bottom: 2px;
  color: #0b0c0c;
}

.c4 {
  margin-bottom: 0;
}

.c5 {
  color: #505a5f;
}

.c6 {
  margin-bottom: 15px;
}

@media print {
  .c2 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c2 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media only screen and (min-width:641px) {
  .c2 {
    margin-bottom: 20px;
  }
}

@media only screen and (min-width:641px) {
  .c1 {
    margin-bottom: 30px;
  }
}

@media only screen and (min-width:641px) {
  .c7 {
    margin-bottom: 0;
  }
}

@media only screen and (min-width:641px) {
  .c8 {
    margin-bottom: 30px;
  }
}

@media only screen and (min-width:641px) {
  .c4 {
    margin-bottom: 0;
  }
}

@media only screen and (min-width:641px) {
  .c6 {
    margin-bottom: 15px;
  }
}

<div
    class="c0 c1"
    data-testid="indicator-search-form"
  >
    <div
      class="c2 c3 c4"
    >
      <p>
        Search by Subject
      </p>
    </div>
    <div
      class="c2 c5 c6"
    >
      <p>
        For example smoking, diabetes prevalence, or a specific indicator ID
      </p>
    </div>
    <div
      class="c7 c8"
    >
      <input
        class="c9"
        data-testid="indicator-search-form-input"
        id="indicator"
        name="indicator"
        style="border-color: #d4351c; border-width: 2px; border-style: solid;"
        title="indicator"
        type="search"
        value=""
      />
      <button
        class="c10"
        data-testid="indicator-search-form-submit"
        title="Search"
        type="submit"
      >
        <svg
          class="c11"
          fill="#ffffff"
          height="100%"
          version="1.1"
          viewBox="0 0 57 57"
          width="100%"
        >
          <title>
            Search
          </title>
          <path
            d="M55.146,51.887L41.588,37.786c3.486-4.144,5.396-9.358,5.396-14.786c0-12.682-10.318-23-23-23s-23,10.318-23,23 s10.318,23,23,23c4.761,0,9.298-1.436,13.177-4.162l13.661,14.208c0.571,0.593,1.339,0.92,2.162,0.92 c0.779,0,1.518-0.297,2.079-0.837C56.255,54.982,56.293,53.08,55.146,51.887z M23.984,6c9.374,0,17,7.626,17,17s-7.626,17-17,17 s-17-7.626-17-17S14.61,6,23.984,6z"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
